packages/langium/src/documents/document.ts

```diff
@@ -4,24 +4,115 @@
-        return TextDocument.update(document, changes, version);
+export interface TextDocumentFactory {
+    fromUri(uri: string): TextDocument;
+}
+
+export class DefaultTextDocumentFactory implements TextDocumentFactory {
+
+    protected readonly languageMetaData: LanguageMetaData;
+
+    constructor(services: LangiumServices) {
+        this.languageMetaData = services.LanguageMetaData;
+    }
+
+    fromUri(uri: string): TextDocument {
+        const content = fs.readFileSync(URI.parse(uri).fsPath, 'utf-8');
+        return TextDocument.create(uri, this.languageMetaData.languageId, 0, content);
+    }
+
+}
+
+export interface LangiumDocumentFactory {
+    fromTextDocument(textDocument: TextDocument): LangiumDocument;
+}
+
+export class DefaultLangiumDocumentFactory implements LangiumDocumentFactory {
+
+    protected readonly parser: LangiumParser;
+
+    constructor(services: LangiumServices) {
+        this.parser = services.parser.LangiumParser;
+    }
+
+    fromTextDocument(textDocument: TextDocument): LangiumDocument {
+        const doc: LangiumDocument = {
+            valid: true,
+            textDocument,
+            parseResult: undefined!
+        };
+        const parseResult = this.parser.parse(doc);
+        doc.parseResult = parseResult;
+        return doc;
     }
-};
+}
+
+export interface LangiumDocuments {
+    readonly all: Stream<LangiumDocument>
+    createOrGetDocument(uri: string): LangiumDocument;
+    invalidateDocument(uri: string): void;
+    invalidateAllDocuments(): void;
+}
+
+export class DefaultLangiumDocuments implements LangiumDocuments {
+
+    protected readonly documentMap: Map<string, LangiumDocument> = new Map();
+    protected readonly textDocuments: TextDocuments<TextDocument>;
+    protected readonly textDocumentFactory: TextDocumentFactory;
+    protected readonly langiumDocumentFactory: LangiumDocumentFactory;
+
+    constructor(services: LangiumServices) {
+        this.textDocuments = services.documents.TextDocuments;
+        this.textDocumentFactory = services.documents.TextDocumentFactory;
+        this.langiumDocumentFactory = services.documents.LangiumDocumentFactory;
+    }
+
+    get all(): Stream<LangiumDocument> {
+        return stream(this.documentMap.values());
+    }
+
+    createOrGetDocument(uri: string): LangiumDocument {
+        let langiumDoc = this.documentMap.get(uri);
+        if (langiumDoc) {
+            return langiumDoc;
+        }
+        let textDoc = this.textDocuments.get(uri);
+        if (!textDoc) {
+            textDoc = this.textDocumentFactory.fromUri(uri);
+        }
+        langiumDoc = this.langiumDocumentFactory.fromTextDocument(textDoc);
+        this.documentMap.set(uri, langiumDoc);
+        return langiumDoc;
+    }
+
+    invalidateDocument(uri: string): void {
+        const langiumDoc = this.documentMap.get(uri);
+        if (langiumDoc) {
+            langiumDoc.valid = false;
```

Comment:
Sorry for the back-and-forth, but seeing this I wonder what the `valid` flag and the "invalidate" methods mean. Do we find a better wording? Maybe
```typescript
langiumDoc.dirty = true;
```
and name the methods `markDirty`, `markAllDirty`?

Reply:
Also, I don't find any code that sets `valid` to `true` again after it has been invalidated. Is that missing in `DocumentBuilder`?

Reply:
Well we throw a document simply away after it has changed, so we never need to set its `valid` state to `true`.

Reply:
Oh, so that's what `valid` means. That's definitely something that should be documented in the LangiumDocument interface. It was not obvious to me.